src/ui/actions/timeline.ts

```diff
@@ -229,6 +243,68 @@ function encodeFocusWindow(focusWindow: TimeStampedPointRange | null) {
   return focusWindow ? encodeObjectToURL(focusWindow) : undefined;
 }
 
+export function step(command: FindTargetCommand): UIThunkAction<Promise<any>> {
+  return async (dispatch, getState, { replayClient }) => {
+    const state = getState();
+    if (getSeekState(state) === "step") {
+      // another call to step() is already in progress, it will execute the enqueued command
+      dispatch(enqueueCommand(command));
+      return;
+    }
+    const focusWindow = replayClient.getCurrentFocusWindow();
+    let point = getExecutionPoint(state);
+    let selectedFrameId = getSelectedFrameId(state);
+    const sourceId = getSelectedSourceId(state);
+    if (!point || !focusWindow) {
+      return;
+    }
+
+    const seekLock = new Object();
+    dispatch(stepping(seekLock));
+    dispatch(enqueueCommand(command));
+
+    let resumeTarget: PauseDescription | undefined;
+    let location: Location | undefined;
+    let nextCommand: FindTargetCommand | undefined = getNextQueuedCommand(getState());
+    while (nextCommand) {
+      dispatch(dequeueCommand());
+
+      resumeTarget = await resumeTargetCache.readAsync(
+        replayClient,
+        nextCommand,
+        point,
+        selectedFrameId,
+        sourceId
+      );
+
+      if (resumeTarget && isPointInRegion(resumeTarget.point, focusWindow)) {
+        point = resumeTarget.point;
+        selectedFrameId = null;
+        location = resumeTarget.frame
+          ? getPreferredLocation(state.sources, resumeTarget.frame)
+          : undefined;
+        if (location) {
+          dispatch(setPreviewPausedLocation(location));
+        } else {
+          dispatch(previewLocationCleared());
+        }
```

Comment:
Would it make sense to only do this for the last command?

I wonder if we're churning things unnecessarily for intermediate states (that we don't expect to be visible for long) in the case where the user has stepped multiple times.

Reply:
I did this for every command because to me it feels better if I can see every step that the debugger made (even if each individual step is only visible for less than a second).